<a href="https://colab.research.google.com/drive/1m__c1gFWmgrVDr9IeMqnQlqVPNwk_MXw" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Importing necessary libraries

In [2]:
import pandas as pd
import numpy as np
from sklearn.datasets import load_iris
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix
from sklearn.linear_model import LogisticRegression
from sklearn import svm
from sklearn.model_selection import train_test_split

In [9]:
# Load the dataset (load remotely, not locally)
irisData = load_iris()
# Output the first 15 rows of the data
irisDf = pd.DataFrame(data= np.c_[irisData['data'], irisData['target']], columns= irisData['feature_names'] + ['target'])
irisDf

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                  5.1               3.5                1.4               0.2   
1                  4.9               3.0                1.4               0.2   
2                  4.7               3.2                1.3               0.2   
3                  4.6               3.1                1.5               0.2   
4                  5.0               3.6                1.4               0.2   
..                 ...               ...                ...               ...   
145                6.7               3.0                5.2               2.3   
146                6.3               2.5                5.0               1.9   
147                6.5               3.0                5.2               2.0   
148                6.2               3.4                5.4               2.3   
149                5.9               3.0                5.1               1.8   

     target  
0       0.0  
1       0.0  
2       0.0  
3       0.0  
4       0.0  
..      ...  
145     2.0  
146     2.0  
147     2.0  
148     2.0  
149     2.0  

[150 rows x 5 columns]

In [10]:

# Display a summary of the table information (number of datapoints, etc.)

irisDf.head(15)


sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                 5.1               3.5                1.4               0.2   
1                 4.9               3.0                1.4               0.2   
2                 4.7               3.2                1.3               0.2   
3                 4.6               3.1                1.5               0.2   
4                 5.0               3.6                1.4               0.2   
5                 5.4               3.9                1.7               0.4   
6                 4.6               3.4                1.4               0.3   
7                 5.0               3.4                1.5               0.2   
8                 4.4               2.9                1.4               0.2   
9                 4.9               3.1                1.5               0.1   
10                5.4               3.7                1.5               0.2   
11                4.8               3.4                1.6               0.2   
12                4.8               3.0                1.4               0.1   
13                4.3               3.0                1.1               0.1   
14                5.8               4.0                1.2               0.2   

    target  
0      0.0  
1      0.0  
2      0.0  
3      0.0  
4      0.0  
5      0.0  
6      0.0  
7      0.0  
8      0.0  
9      0.0  
10     0.0  
11     0.0  
12     0.0  
13     0.0  
14     0.0

In [11]:
irisDf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
 4   target             150 non-null    float64
dtypes: float64(5)
memory usage: 6.0 KB


# About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?

Dataset consists of 150 rows with no null values, there are total of 5 columns: <br> sepal length, sepal width, petal length  petal width, and target for Iris setosa, r Iris versicolor and Iris virginica.

Features: 

sepal length (cm)<br>
sepal width (cm)<br>
petal length (cm)<br>
petal width (cm)<br>

Label: target

# Part 2: Split the dataset into train and test

In [14]:
# Take the dataset and split it into our features (X) and label (y)
X = irisData.data
y = irisData.target


In [69]:
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=.90, random_state=40)

# Part 3: Logistic Regression

In [71]:
# i. Use sklearn to train a LogisticRegression model on the training set

regression = LogisticRegression()
regression.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class

y_pred = regression.predict(X_test)
y_pred_prob = regression.predict_proba(X_test)
classes = load_iris().target_names
for class_name, proba in zip(classes, y_pred_prob):
    index = np.argmax(proba)
    print(f'predicted class : {class_name} ; confidence : {proba[index]:.2%}')
    
# iii. Report on the score for Logistic regression model, what does the score measure?
score = regression.score(X_test, y_test)
print("Score is", score)

confusion_matrix(y_test,regression.predict(X_test))


predicted class : setosa ; confidence : 98.50%
predicted class : versicolor ; confidence : 74.04%
predicted class : virginica ; confidence : 93.86%
Score is 1.0


array([[4, 0, 0],
       [0, 7, 0],
       [0, 0, 4]], dtype=int64)

We can see that the score for this model stands on 100%. <br>
The confusion matrix showes setosa we have 4 correct predicted setosa and 0 false valuse. Conclusion: very high accuracy. <br>
For versicolor we got 7 predicted correclty and 0 for falsely predictions (all are zeros). <br>
For virginicia again we got 4 correct predictions with no false predictions, which again tells up we have very high accuracy. 


In [63]:
# iv. Extract the coefficents and intercepts for the boundary line(s)
intercept = regression.intercept_
coeff = regression.coef_.T
print("intercept:", intercept, "\n\ncoeff:" , coeff)

intercept: [  9.28732948   1.71988977 -11.00721925] 

coeff: [[-0.37999813  0.59263249 -0.21263436]
 [ 0.96207161 -0.30953845 -0.65253316]
 [-2.4418702  -0.2274171   2.66928729]
 [-1.05257041 -0.85494491  1.90751532]]


# Part 4: Support Vector Machine

In [64]:
# i. Use sklearn to train a Support Vector Classifier on the training set
svc = svm.SVC(probability=True)
svc.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class

y_pred_svc = svc.predict(X_test)
y_pred_prob_svc = svc.predict_proba(X_test)
for class_name, proba in zip(classes, y_pred_prob_svc):
    index = np.argmax(proba)
    print("Predicted class :", class_name ,"\nConfidence :", proba[index], "\n")

# iii. Report on the score for the SVM, what does the score measure?

score = svc.score(X_test, y_test)
print("Socre:", score)

print(confusion_matrix(y_test,svc.predict(X_test)))


Predicted class : setosa 
Confidence : 0.9736340340675134 

Predicted class : versicolor 
Confidence : 0.8648326618353246 

Predicted class : virginica 
Confidence : 0.9815437676580743 

Socre: 1.0
[[4 0 0]
 [0 7 0]
 [0 0 4]]


Once again, we see that the coeff matrix produces high accuracy, as the predicted vs. the actual classes match with zeros false predictions accordingly.  <br> Score here also stands on 100%.

# Part 5: Neural Network

In [60]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
mlp = MLPClassifier(solver='adam',learning_rate_init=0.01, hidden_layer_sizes=10, max_iter=600,random_state=1)
mlp.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
y_pred_mlp = mlp.predict(X_test)
y_pred_prob_mlp = mlp.predict_proba(X_test)

for class_name, proba in zip(classes, y_pred_prob_mlp):
    index = np.argmax(proba)
    print("predicted class:", class_name ,"; confidence:" ,proba[index])

    # iii. Report on the score for the Neural Network, what does the score measure?
score = mlp.score(X_test, y_test)
print("accuracy score: ",score)

# iv: Experiment with different options for the neural network, report on your best configuration (the highest score I was able to achieve was 0.8666)

predicted class: setosa ; confidence: 0.9850048305800763
predicted class: versicolor ; confidence: 0.999926133019437
predicted class: virginica ; confidence: 0.9999909734444146
accuracy score:  0.9333333333333333


For this model the score here produces 93% percision.

# Part 6: K-Nearest Neighbors

In [59]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier

knn = KNeighborsClassifier(n_neighbors=8)
knn.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class

y_pred_knn = knn.predict(X_test)
y_pred_prob_knn = knn.predict_proba(X_test)
for class_name, proba in zip(classes, y_pred_prob_knn):
    index = np.argmax(proba)
    print("predicted class:", class_name," ; confidence:", proba[index])

    # iii. Report on the score for kNN, what does the score measure?

    score = knn.score(X_test, y_test)
print("accuracy:",format(score))

predicted class: setosa  ; confidence: 1.0
predicted class: versicolor  ; confidence: 1.0
predicted class: virginica  ; confidence: 1.0
accuracy: 1.0


For this model the score here produces 100% percision.

# Part 7: Conclusions and takeaways


In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

Answer:

Four different types of classifiers were used to solve this assignment:
Logistic Regression
Support Vector Machine
Neural Network
K-Nearest Neighbors

It quite surprised me that all 4 model except the neural network the models produced 100% accuracy.
It is interesting to see they all produced 100% or very closed to a perfect precision.
I am trying to think perhaps it is due to a very precise data set quality that allowed each model to produce clear distinction between the 3 different classes.
